src/main/java/io/cdap/plugin/http/sink/batch/MessageBuffer.java

```diff
@@ -0,0 +1,232 @@
+        contentType = "text/plain";
+        break;
+      default:
+        throw new IllegalArgumentException("Invalid message format: " + messageFormat);
+    }
+  }
+
+  /**
+   * Adds a record to the buffer.
+   *
+   * @param record The record to be added.
+   */
+  public void add(StructuredRecord record) {
+    buffer.add(record);
+  }
+
+  /**
+   * Clears the buffer.
+   */
+  public void clear() {
+    buffer.clear();
+  }
+
+  /**
+   * Returns the size of the buffer.
+   */
+  public int size() {
+    return buffer.size();
+  }
+
+  /**
+   * Returns whether the buffer is empty.
+   */
+  public boolean isEmpty() {
+    return buffer.isEmpty();
+  }
+
+  /**
+   * Returns the content type of the message.
+   */
+  public String getContentType() {
+    return contentType;
+  }
+
+  /**
+   * Converts the buffer to the appropriate format and returns the message.
+   */
+  public String getMessage() throws IOException {
+    return messageFormatter.apply(buffer);
+  }
+
+  private String formatAsJson(List<StructuredRecord> buffer) {
+    try {
+      return formatAsJsonInternal(buffer);
+    } catch (IOException e) {
+      throw new IllegalStateException("Error formatting JSON message. Reason: " + e.getMessage(), e);
+    }
+  }
+
+  private String formatAsJsonInternal(List<StructuredRecord> buffer) throws IOException {
+    boolean useJsonBatchKey = !Strings.isNullOrEmpty(jsonBatchKey);
+    if (!shouldWriteJsonAsArray || !useJsonBatchKey) {
+      return getBufferAsJsonList();
+    }
+    // A new StructuredRecord is created with the jsonBatchKey as the
+    // field name and the array of records as the value
+    Schema bufferRecordSchema = buffer.get(0).getSchema();
+    Schema bufferRecordArraySchema = Schema.arrayOf(bufferRecordSchema);
+    Schema wrappedMessageSchema = Schema.recordOf("wrapper",
+            Schema.Field.of(jsonBatchKey, bufferRecordArraySchema));
+    StructuredRecord wrappedMessageRecord = StructuredRecord.builder(wrappedMessageSchema)
+            .set(jsonBatchKey, buffer).build();
+    return StructuredRecordStringConverter.toJsonString(wrappedMessageRecord);
+  }
+
+  private String formatAsForm(List<StructuredRecord> buffer) {
+    return buffer.stream()
+            .map(this::createFormMessage)
+            .collect(Collectors.joining(delimiterForMessages));
+  }
+
+  private String formatAsCustom(List<StructuredRecord> buffer) {
+    return buffer.stream()
+            .map(this::createCustomMessage)
+            .collect(Collectors.joining(delimiterForMessages));
+  }
+
+  private String getBufferAsJsonList() throws IOException {
+    StringBuilder sb = new StringBuilder();
+    int size = buffer.size();
+    String delimiter = shouldWriteJsonAsArray ? "," : delimiterForMessages;
+    if (shouldWriteJsonAsArray) {
+      sb.append("[");
+    }
+    for (StructuredRecord record : buffer) {
+      sb.append(StructuredRecordStringConverter.toJsonString(record));
+      if (--size > 0) {
```

Comment:
nit: since we're using a StringBuilder, it's more efficient to remove the last character at the end of the loop (assuming at least 1 element in the buffer) rather than doing this check every time through the loop.

Reply:
```java
if (!buffer.isEmpty()) {
  sb.setLength(sb.length() - delimiter.length());
}
 ```